test/clj/rems/application/test_model.clj

```diff
@@ -848,10 +855,49 @@
                                                                  :application/events events
                                                                  :application/members #{}
                                                                  :application/past-members #{{:userid "member"}}})]
-                                (is (= expected-application (apply-events events)))))))))))))))))))
+                                (is (= expected-application (apply-events events)))))))))))))))))
+
+    (testing "generate report: permissions by role and state"
+      (let [apps @sample-applications
+            state-role-permissions (fn [app]
+                                     (map (fn [[role permissions]]
+                                            {:state (:application/state app)
+                                             :role role
+                                             :permissions permissions})
+                                          (:rems.permissions/role-permissions app)))
+            data (mapcat state-role-permissions apps)
+            states (->> data (map :state) distinct) ; keep states in the order they appear in the tests
+            roles (->> data (map :role) distinct sort)]
+        (->> (hiccup/html
+              [:table {:border 1}
+               [:tr
+                [:th "State \\ Role"]
+                (for [role roles]
+                  [:th (str role)])]
+               (for [state states]
+                 [:tr
+                  [:th {:style "text-align: left; vertical-align: top"}
+                   (str state)]
+                  (for [role roles]
+                    (let [perm-sets (->> data
+                                         (filter #(= state (:state %)))
+                                         (filter #(= role (:role %)))
+                                         (map :permissions))
+                          all-perms (apply set/union perm-sets)
+                          ;; the states and permissions are not guaranteed to have a 1:1 mapping,
+                          ;; so we separate conditional permissions from those that are always there
+                          always-perms (if (empty? perm-sets)
+                                         #{}
+                                         (apply set/intersection perm-sets))
+                          sometimes-perms (set/difference all-perms always-perms)]
+                      [:td {:style "vertical-align: top"}
+                       (for [perm (sort always-perms)]
+                         [:div (str perm)])
+                       (for [perm (sort sometimes-perms)]
+                         [:div [:i "(" (str perm) ")"]])]))])])
+             (spit "docs/application-permissions.html"))))))
```

Comment:
I'm not sure I like tests that modify checked-in files, but I guess this is a nice way to make sure the file is up-to-date.